extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java

```diff
@@ -0,0 +1,126 @@
+import io.ap4k.processor.SimpleFileWriter;
+import io.ap4k.project.Project;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.ApplicationInfoBuildItem;
+import io.quarkus.deployment.builditem.FeatureBuildItem;
+import io.quarkus.deployment.builditem.GeneratedResourceBuildItem;
+import io.quarkus.kubernetes.spi.KubernetesPortBuildItem;
+
+class KubernetesProcessor {
+
+    @Inject
+    BuildProducer<GeneratedResourceBuildItem> generatedResourceProducer;
+
+    @Inject
+    BuildProducer<FeatureBuildItem> featureProducer;
+
+    @BuildStep
+    public void build(ApplicationInfoBuildItem applicationInfo,
+            KubernetesConfig kubernetesConfig,
+            List<KubernetesPortBuildItem> kubernetesPortBuildItems) {
+
+        if (kubernetesPortBuildItems.isEmpty()) {
+            return;
+        }
+
+        // The resources that ap4k's execution will result in, will later-on be written
+        // by quarkus in the 'wiring-classes' directory
+        // The location is needed in order to properly support s2i build triggering
+
+        // by passing false to SimpleFileWriter, we ensure that no files are actually written during this phase
+        final Path root;
+        try {
+            root = Files.createTempDirectory("quarkus-kubernetes");
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to setup environment for generating Kubernetes resources", e);
+        }
+
+        final SessionWriter sessionWriter = new SimpleFileWriter(root, false);
+        sessionWriter.setProject(new Project());
+        final Session session = Session.getSession();
+        session.setWriter(sessionWriter);
+
+        final Map<String, Integer> ports = verifyPorts(kubernetesPortBuildItems);
+        enableKubernetes(applicationInfo, kubernetesConfig, ports);
+
+        // write the generated resources to the filesystem
+        final Map<String, String> generatedResourcesMap = session.close();
+        for (String generatedResourceFullPath : generatedResourcesMap.keySet()) {
+            generatedResourceProducer.produce(
+                    new GeneratedResourceBuildItem(
+                            // we need to make sure we are only passing the relative path to the build item
+                            generatedResourceFullPath.replace(root.toAbsolutePath() + "/", "META-INF/ap4k/"),
+                            generatedResourcesMap.get(generatedResourceFullPath).getBytes()));
+        }
+
+        featureProducer.produce(new FeatureBuildItem(FeatureBuildItem.KUBERNETES));
+    }
+
+    private Map<String, Integer> verifyPorts(List<KubernetesPortBuildItem> kubernetesPortBuildItems) {
+        final Map<String, Integer> result = new HashMap<>();
+        final Set<Integer> usedPorts = new HashSet<>();
+        for (KubernetesPortBuildItem entry : kubernetesPortBuildItems) {
```

Comment:
Really like the fact that we perform this verification!

Reply:
The more build time errors the better :)